core/services/sleeper_task.go

```diff
@@ -47,84 +28,60 @@ type sleeperTask struct {
 // WakeUp does not block.
 //
 func NewSleeperTask(worker Worker) SleeperTask {
-	return &sleeperTask{
-		worker: worker,
-		waker:  make(chan struct{}, 1),
-		closer: make(chan struct{}, 1),
-		closed: make(chan struct{}, 1),
-		state:  unstarted,
-		mutex:  sync.Mutex{},
+	s := &sleeperTask{
+		worker:      worker,
+		chQueue:     make(chan struct{}, 1),
+		chQueueDone: make(chan struct{}),
+		chStopped:   make(chan struct{}),
 	}
-}
 
-// Start begins the SleeperTask
-// Calling start on a started task will return an error
-func (s *sleeperTask) Start() error {
-	s.mutex.Lock()
-	defer s.mutex.Unlock()
-	switch s.state {
-	case stopped:
-		return errors.New("cannot start a sleeper task that has been stopped")
-	case started:
-		return errors.New("sleeper task is already started")
-	}
-	s.state = started
 	go s.workerLoop()
-	return nil
+
+	return s
 }
 
-// Stop stops the SleeperTask
-// It can only be called once
-// It will block until all tasks have completed, or the timeout is reached
-// Stopped tasks cannot be restarted
+// Stop stops the SleeperTask.  It never returns an error.  Its error return
+// exists so as to satisfy other interfaces.
 func (s *sleeperTask) Stop() error {
-	s.mutex.Lock()
-	if s.state == stopped {
-		s.mutex.Unlock()
-		return errors.New("sleeper task is already stopped")
-	}
-	s.state = stopped
-	s.mutex.Unlock()
-
-	s.closer <- struct{}{}
-	// NOTE: Closing the channels will cause the rogue task to panic if it does eventually complete
-	defer close(s.waker)
-	defer close(s.closer)
-	defer close(s.closed)
 	select {
-	case <-s.closed:
-		return nil
-	case <-time.After(stopWaitTime):
-		return fmt.Errorf("task did not complete within %s", stopWaitTime.String())
+	case <-s.chStopped:
+		panic("already stopped")
```

Comment:
why not just return an error?

Reply:
It's a programming error to call Stop twice (meaning that it's not the sort of thing you should ever try to recover from in a running node).  A panic is the canonical way to address that.